Add map update tracking to MapCache

PunkPun opened the pull request:
fixes #20012

I also update LobbyLogic to use this new logic, it reduces maintenance overhead

PunkPun committed:
Add map update tracking to MapCache and fix crash when restarting a game

pchote commented:
Are we able to use this to fix #19988 too?

PunkPun commented:
> Are we able to use this to fix #19988 too?

I will just update #19990 to depend on this pr. Really it is already ready to merge, all I can do is just add smart updating to it (which just became very easy with this pr)

pchote commented:
I meant that applying the approach from `RestartGame` to `ServerCreationLogic.CreateAndJoin` should fix https://github.com/OpenRA/OpenRA/pull/19990/files#r858018992 and make that PR unnecessary(?)

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -232,6 +232,21 @@ public static void RestartGame()
 			// Reseed the RNG so this isn't an exact repeat of the last game
 			lobbyInfo.GlobalSettings.RandomSeed = CosmeticRandom.Next();
 
+			// Update MapCache
```
Can we please change this to say something like
```
// Note: the map may have been changed on disk outside the game, changing its UID.
// Use the updated UID if we have tracked the update instead of failing.
```
?

PunkPun replied:
sure

pchote commented on OpenRA.Game/Map/MapCache.cs:
```diff
@@ -41,11 +41,8 @@ public sealed class MapCache : IEnumerable<MapPreview>, IDisposable
-		/// </summary>
-		public event Action<string, string> MapUpdated = (oldUID, newUID) => { };
+		readonly Dictionary<string, string> mapUpdates = new Dictionary<string, string>();
+		public string LastUpdatedMap = null;
```
Are we able to make this `public string LastUpdatedMap { get; private set; }` to ensure that others can't modify it?

PunkPun replied:
sure

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -232,6 +232,21 @@ public static void RestartGame()
 			lobbyInfo.GlobalSettings.RandomSeed = CosmeticRandom.Next();
 
+			// Update MapCache
+			if (ModData.MapCache[lobbyInfo.GlobalSettings.Map].Status != MapStatus.Available)
```
This check doesn't appear to be needed (and isn't used in `LobbyLogic`)?

PunkPun replied:
it is used when you restart the game, its part of the code that fixes the #20012 crash. It both updates MapCache and checks if the selected map is not outdated

pchote replied:
I mean that we can run the body of this unconditionally, since `GetUpdatedMap` will also update the `MapCache` and return the original UID unchanged if it still exists.

PunkPun replied:
you are correct, I'll remove the check

PunkPun commented:
> I meant that applying the approach from `RestartGame` to `ServerCreationLogic.CreateAndJoin` should fix https://github.com/OpenRA/OpenRA/pull/19990/files#r858018992 and make that PR unnecessary(?)

I disagree on such an approach. As I mentioned in that PR I find my approach more intuitive and user friendly, it mimics the behaviour of `LobbyLogic`, and to some extent, `MissionBrowser`

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -193,17 +189,14 @@ void ConnectionStateChanged(OrderManager om, string password, NetworkConnection
 					Ui.OpenWindow("MAPCHOOSER_PANEL", new WidgetArgs()
 					{
-						{ "initialMap", lastUpdatedUid ?? map.Uid },
+						{ "initialMap", modData.MapCache.LastUpdatedMap ?? map.Uid },
```
Can `LastUpdatedMap` ever be `null`? It is set in `LoadMap` (which would crash before setting it to a null value), and that runs already during mod initialisation. 🤔

abcdefg30 replied:
Confirmed that on bleed the map selected in the lobby is also selected when opening the map chooser, but on this PR always the map loaded last when initialising the mod is selected. (Until you change a map while the game is running and that becomes the permanently selected map.)

PunkPun replied:
Yes, this is a regression I though of earlier today. I will fix it in the near future

PunkPun replied:
fixed it

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -59,9 +59,8 @@ enum PanelType { Players, Options, Music, Servers, Kick, ForceStart }
-		string newMapUid;
-		string lastUpdatedUid;
+
+		string lastUpdatedMap = null;
```
I'm not really sure what name to use here

anvilvapre commented:
I notice that the map tracker maintains a `FileSystemWatcher` for each map directory and all files in it. That is usefull during map editing. But is it needed in game? Wouldn't it be better to only enable this functionality with a start up flag or setting checkbox?

How many maps does a player have in the map directory? File watching can be slow on slow old systems with many files to watch. (Can you even watch for filename changes/moves on Linux?)

`LastUpdatedMap` how about `LastExternallyModifiedMap`. As i read the code, when you restart a game it will reload any other map that was last modified on disk? So this could also be a different map? I guess this is what you want as editor.

Map cache code flow is somewhat difficult to read. The directory trackers call `LoadMap` as well as `MapCache.LoadMap` does. Seems cleaner if `MapCache` would receive events from the trackers and reload maps by itself based on the event? Both classes/object call eachother in each direction.

With now `mapUpdates` it seems updates are coming from multiple sides? I.e. you add an additional layer to track the same changes?

Seems incorrect that `mapUpdates` is only added to never removed?

Entries for old maps/hashes remain in the cache? This might be ok - assuming there won't be many updates over time.

PunkPun commented:
> I notice that the map tracker maintains a `FileSystemWatcher` for each map directory and all files in it. That is usefull during map editing. But is it needed in game? Wouldn't it be better to only enable this functionality with a start up flag or setting checkbox?
>
> How many maps does a player have in the map directory? File watching can be slow on slow old systems with many files to watch. (Can you even watch for filename changes/moves on Linux?

Afaik it works on linux, pretty sure someone had tested that. About performance I dunno, my code shouldn't have any impact, but I have no idea about `FileSystemWatcher`. Tbh I never even considered it, just followed pchotes suggestion to use it

> Seems incorrect that `mapUpdates` is only added to never removed?
> 
> Entries for old maps/hashes remain in the cache? This might be ok - assuming there won't be many updates over time.

That's how the system worked before I implemented updates. Outdated maps get disabled, but not deleted from memory. Because of that I implemented a system that only adds a new map when necessary, as I don't want to clog up memory.

I'll take a deeper look into your other suggestions later, when I get back home tomorrow.

PunkPun commented:
> `LastUpdatedMap` how about `LastExternallyModifiedMap`. As i read the code, when you restart a game it will reload any other map that was last modified on disk? So this could also be a different map? I guess this is what you want as editor.

I'm renaming to `LastModifiedMap`. I cannot name it external as map editor is also tracked. 
This does not track anything on game restart as the game doesn't keep track of files without being turned on. Rather files are tracked during runtime and MapCache is updated whenever a map is requested by any part of the code, during runtime

> Map cache code flow is somewhat difficult to read. The directory trackers call `LoadMap` as well as `MapCache.LoadMap` does. Seems cleaner if `MapCache` would receive events from the trackers and reload maps by itself based on the event? Both classes/object call eachother in each direction.

What you are basically suggesting is moving `MapDirectoryTracker` class into `MapCache`. We can't use events without moving basically all tracking logic to `MapCache`. We could of course do that and stack `MapCache` even harder. I wouldn't oppose it but I don't really see point so it won't be me doing the refactoring.

> With now `mapUpdates` it seems updates are coming from multiple sides? I.e. you add an additional layer to track the same changes?

I'm not really sure what you mean. Values get added to it only from `LoadMap` method

PunkPun committed:
Update LobbyLogic to use the new MapCache map tracking

abcdefg30 reviewed (approved):
Fix confirmed.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/220ca7f4740472c9f36bee7741d7e5c8765c5cf8...bdbc00046461480bb6fdb59c6f533f5077090739)
